### PR TITLE
[TTS] Create script for processing TTS training audio

### PR DESCRIPTION
**Pull request by rlangman**
Signed-off-by: Ryan <rlangman@nvidia.com>

# What does this PR do ?

Add a script for doing audio preprocessing for TTS training.

**Collection**: [TTS]

# Changelog 
- Created a "data" directory under nemo.collections.tts following the same convention as the ASR & NLP code.
- Added data_util class for generic operations like processing training manifest. Not sure of volume normalization should be there, but currently have no better place to put it.
- Added silence trimming implementations using energy and VAD based approaches.
- Added audio_preprocessing script for running the data processing logic over a training manifest, with example config files.

# Usage

```
HYDRA_FULL_ERROR=1 python /home/rlangman/Code/NeMo/scripts/dataset_processing/tts/audio_processing/preprocess_audio.py \
    --config-path=/home/rlangman/Code/NeMo/scripts/dataset_processing/tts/audio_processing/config \
	--config-name=preprocessing.yaml \
	data_base_dir="/home/rlangman/Data/nemo_processing" \
	config.num_workers=2 \
	config.max_entries=100 \
	trim=vad \
	trim.device=cuda
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

# Additional Information

I manually tested the silence trimming using the openSLR Spanish data. For the 2 modes the dataset size decreases from:

Energy: 35 hours -> 28.8 hours
VAD: 35 hours -> 22.6 hours

This is because lots of the audio has noise in the long silences at the start and end of files. The VAD can differentiate the noise from speech while energy cannot.

Doing trimming on my workstation with the ~24k utterance dataset takes about 5 minutes with energy and 30 minutes with VAD.

More automated testing might be possible if we add CI tests that use actual audio, but I would leave that for a future PR as we do not yet have guidelines for TTS CI tests.

### CODE REVIEW

**Review by redoctopus — COMMENTED** [review]
Thanks for adding this!

File: nemo/collections/tts/data/audio_trimming.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+            start_i = i
+            break
+
+    # Iterate backwards over the utterance until we find the last frame_threshold consecutive speech frames.
+    end_i = None
+    for i in range(num_frames, 0, -1):
+        low_i = max(0, i - frame_threshold)
+        if all(is_speech[low_i:i]):
+            end_i = i
+            break
```
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/data/audio_trimming.py:256]: The min/max here might not make sense (I wrote it originally with the check sum() > threshold rather than all()). I think I should stop the iteration when the range becomes less than the threshold size.

File: nemo/collections/tts/data/audio_trimming.py [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+        # [num_frames]
+        speech_probs = probs[:, 1]
+        speech_frames = speech_probs >= self.vad_threshold
+
+        return speech_frames
+
+    def _scale_sample_indices(self, start_sample_i: int, end_sample_i: int, sample_rate: int) -> Tuple[int, int]:
+        sample_rate_ratio = sample_rate / self.vad_sample_rate
+        start_sample_i = sample_rate_ratio * start_sample_i
+        end_sample_i = sample_rate_ratio * end_sample_i
```
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/data/audio_trimming.py:188]: pad_sample_indices() called after this is what handles the bounding and integer conversion. I could also bound the min/max and convert the final output to integer within this function too if it makes things clearer.

File: scripts/dataset_processing/tts/audio_processing/config/preprocessing.yaml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  _target_: scripts.dataset_processing.tts.audio_processing.preprocess_audio.AudioPreprocessingConfig
+  input_manifest: ${data_base_dir}/manifest.json
+  output_manifest: ${data_base_dir}/manifest_processed.json
+  output_dir: ${data_base_dir}/audio_processed
+  num_workers: -1
+  max_entries: 0
+  output_sample_rate: 0
+  volume_level: 0.95
+  min_duration: 0.5
+  max_duration: 10.0
```
[the remark this answers is not in the document]
Comment (rlangman) [reply scripts/dataset_processing/tts/audio_processing/config/preprocessing.yaml:18]: In the end these are just examples that people will want to tweak depending on their dataset.

An upper bound of 10 seconds is a rule of thumb. I don't remember where I heard it originally, but I heard the same thing from Boris/Evelina months ago when we were talking about creating new HiFi-TTS like datasets. Its mainly done for computation reasons; the compute time and memory for a training batch is primarily dependent on the longest sequence in the batch. To use utterances much longer than 10 seconds you typically implement duration based batching (eg. there are a lot of papers that claim they use batch sizes of 10k or 20k audio frames, instead of fixed utterance count). A reason it comes up often is because TTS datasets are often derived from ASR datasets, and ASR likes training with utterances up to 20 seconds long.

The lower bound is more arbitrary. I would be OK with setting it to 0.1 for consistency. The reason I would set it higher than 0.1 here is to help identify/remove edge cases where the trimming ends up removing too much of the utterance. Alternatively I considered storing the original and trimmed durations in the output manifest and leaving the duration filtering/validation/analysis as a second step/script (the filtering could be done with a simple JQ command https://stedolan.github.io/jq/), but thought that was too complicated for an initial implementation.

File: nemo/collections/tts/data/audio_trimming.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (redoctopus) [reply nemo/collections/tts/data/audio_trimming.py:188]: Ah gotcha, I see it. Either is fine, though if it's kept as-is can you put in a note for the future in case anyone wants to call it by itself for whatever reason?

File: scripts/dataset_processing/tts/audio_processing/config/preprocessing.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (redoctopus) [reply scripts/dataset_processing/tts/audio_processing/config/preprocessing.yaml:18]: In general I'd be worried that people will blindly take whatever parameters are set in the example as defaults :P The reasoning is good and these seem fine, though.

File: nemo/collections/tts/data/audio_trimming.py [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+        probs = torch.softmax(log_probs, dim=-1)
+        probs = probs.cpu().detach().numpy()
+        # [num_frames]
+        speech_probs = probs[:, 1]
+        speech_frames = speech_probs >= self.vad_threshold
+
+        return speech_frames
+
+    def _scale_sample_indices(self, start_sample_i: int, end_sample_i: int, sample_rate: int) -> Tuple[int, int]:
+        sample_rate_ratio = sample_rate / self.vad_sample_rate
```
Comment (XuesongYang) [inline nemo/collections/tts/data/audio_trimming.py:186]: add `int()` to enforce return type hint as `Tuple[int, int]`. I would add try...except just in case zero-division.

File: nemo/collections/tts/data/audio_trimming.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (XuesongYang) [reply nemo/collections/tts/data/audio_trimming.py:188]: +1 with above suggestion.

File: scripts/dataset_processing/tts/audio_processing/preprocess_audio.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+and duration filtering.
+
+These can be done separately through multiple executions of the script, or all at once to avoid saving
+too many copies of the same audio.
+
+Most of these can also be done by the TTS data loader at training time, but doing them ahead of time
+lets us implement more complex processing, validate the corectness of the output, and save on compute time.
+
+$ HYDRA_FULL_ERROR=1 python /home/rlangman/Code/NeMo/scripts/dataset_processing/tts/audio_processing/preprocess_audio.py \
+    --config-path=/home/rlangman/Code/NeMo/scripts/dataset_processing/tts/audio_processing/config \
```
Comment (XuesongYang) [inline scripts/dataset_processing/tts/audio_processing/preprocess_audio.py:28]: nit: `/home/rlangman/Code/NeMo` -> `<your_nemo_repo_root>`.

File: scripts/dataset_processing/tts/audio_processing/config/preprocessing.yaml [hunk 6]
```diff
@@ -0,0 +1,19 @@
+name: "preprocessing"
+
+data_base_dir: "/home"
```
Comment (XuesongYang) [inline scripts/dataset_processing/tts/audio_processing/config/preprocessing.yaml:3]: shall we leave it `???` or keep as-is?

File: scripts/dataset_processing/tts/audio_processing/preprocess_audio.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+) -> Tuple[dict, float, float]:
+    audio_filepath = Path(entry["audio_filepath"])
+    rel_audio_path = audio_filepath.relative_to(base_dir)
+    input_path = os.path.join(base_dir, rel_audio_path)
+    output_path = os.path.join(output_dir, rel_audio_path)
+
+    audio, sample_rate = librosa.load(input_path, sr=None)
+
+    if audio_trimmer is not None:
+        audio, start_i, end_i = audio_trimmer.trim_audio(audio=audio, sample_rate=sample_rate, audio_id=audio_filepath)
```
Comment (XuesongYang) [inline scripts/dataset_processing/tts/audio_processing/preprocess_audio.py:92]: convert to string: `audio-id` is `str` but `audio_filepath` is `Path`.

File: scripts/dataset_processing/tts/audio_processing/preprocess_audio.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+    if output_sample_rate is not None:
+        audio = librosa.resample(y=audio, orig_sr=sample_rate, target_sr=output_sample_rate)
+        sample_rate = output_sample_rate
+
+    if volume_level:
+        audio = normalize_volume(audio, volume_level=volume_level)
+
+    sf.write(file=output_path, data=audio, samplerate=sample_rate)
+
+    original_duration = librosa.get_duration(filename=audio_filepath)
```
Comment (XuesongYang) [inline scripts/dataset_processing/tts/audio_processing/preprocess_audio.py:103]: convert to string: `filename` is `str` but `audio_filepath` is `Path`.

File: scripts/dataset_processing/tts/audio_processing/preprocess_audio.py (hunk 7 above)
Comment (XuesongYang) [inline scripts/dataset_processing/tts/audio_processing/preprocess_audio.py:92]: convert to string: `audio-id` is `str` but `audio_filepath` is `Path`.

File: scripts/dataset_processing/tts/audio_processing/config/preprocessing.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (XuesongYang) [reply scripts/dataset_processing/tts/audio_processing/config/preprocessing.yaml:18]: thanks for explanation. how about make it consistent with other yamls, and leave a comment to user?

**Review by XuesongYang — COMMENTED** [review]
LGTM. Added some nitpicks and discussion. 

It seems that this PR support either energy-based or vad-based trimming solution. For entergy-based method, the classic way is using two measures of energy and  zero-crossing rate (see ref https://web.ece.ucsb.edu/Faculty/Rabiner/ece259/Reprints/078_endpoint%20detection.pdf). We are good to ship energy only method, but we can add ZCR in future.

another thought, is it possible to make a hybrid solution with both energy and VAD?

File: nemo/collections/tts/data/audio_trimming.py [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+                 considered silence.
+               ref_amplitude: Amplitude threshold for classifying speech versus silence.
+               frame_threshold: Start and end of speech will be detected where there are at least frame_threshold
+                 consecutive audio frames classified as speech. Setting this value higher is more robust to
+                 false-positives (silence detected as speech), but setting it too high may result in very short
+                 speech segments being cut out from the audio.
+               frame_length: Length of audio frames to use when doing speech detection. This does not need to match
+                 the frame_length used any other part of the code or model.
+               frame_step: Stride of audio frames to use when doing speech detection. This does not need to match
+                 the frame_step used any other part of the code or model.
```
[the remark this answers is not in the document]
Comment (redoctopus) [reply nemo/collections/tts/data/audio_trimming.py:68]: I'm in favor of that. It's pretty explicit here in the docstring that they can be different, but maybe we can call it `trim_win_length` and `trim_hop_length`?

**Comment by rlangman** [comment]
> It seems that this PR support either energy-based or vad-based trimming solution. For entergy-based method, the classic way > is using two measures of energy and zero-crossing rate (see ref https://web.ece.ucsb.edu/Faculty/Rabiner/ece259/Reprints /078_endpoint%20detection.pdf). We are good to ship energy only method, but we can add ZCR in future.

> another thought, is it possible to make a hybrid solution with both energy and VAD?

The intention is to merge a prototype/skeleton implementation here and then improve on the algorithm(s) over time. ZCR could definitely be useful.

A hybrid solution with both energy and VAD also sounds like a good idea. I will need to think about how to refactor the code to support such a mode. For those 2 you can simply take the boolean speech frame array and do an AND with the two outputs, but that might not be as applicable for other approaches.

**Comment by redoctopus** [comment]
@rlangman what's the status of this PR?

**Comment by rlangman** [comment]
> @rlangman what's the status of this PR?

I will update the PR next week. In addition to addressing the PR comments, I talked with Fei and am seeing if I can make the VAD config more consistent with what it was trained on and potentially reuse some of the code from the VAD post-processor.

File: nemo/collections/tts/data/audio_trimming.py [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+        audio_signal = torch.tensor(audio_frames, dtype=torch.float32, device=self.device)
+        # [1]
+        audio_signal_len = torch.tensor(num_frames * [self.frame_length], dtype=torch.int32, device=self.device)
+
+        # VAD outputs 2 values for each audio frame with logits indicating the likelihood that
+        # each frame is non-speech or speech, respectively.
+        # [num_frames, 2]
+        log_probs = self.vad_model(input_signal=audio_signal, input_signal_length=audio_signal_len)
+        probs = torch.softmax(log_probs, dim=-1)
+        probs = probs.cpu().detach().numpy()
```
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/data/audio_trimming.py:178]: Since VAD can run on GPU, transferring to CPU is needed to process it as numpy afterwards. Technically I might be able to transfer it after the next 2 lines.

I will swap the order of the detach() and cpu().

File: nemo/collections/tts/data/audio_trimming.py [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+
+        start_i, end_i = pad_sample_indices(
+            start_sample_i=start_i,
+            end_sample_i=end_i,
+            max_sample=audio.shape[0],
+            sample_rate=sample_rate,
+            pad_seconds=self.pad_seconds,
+        )
+
+        trimmed_audio = audio[start_i:end_i]
```
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/data/audio_trimming.py:223]: The +1 is implicitly inside the "get_start_and_end_of_speech" logic (as can be seen in the unit tests for the function). Its a bit hard to follow, but otherwise the +1 needs to be done at the frame level and before the padding rather than in this line at the sample level. I will see if I can make it clearer.

File: nemo/collections/tts/data/data_utils.py [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+        output_f.writelines(output_lines)
+
+
+def normalize_volume(audio: np.array, volume_level: float):
+    """Apply peak normalization to the input audio.
+    """
+    if not (0.0 <= volume_level <= 1.0):
+        raise ValueError(f"Volume must be in range [0.0, 1.0], received {volume_level}")
+
+    return volume_level * (audio / np.max(np.abs(audio)))
```
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/data/data_utils.py:44]: I think all-zero audio is technically silence, so volume change should be a no-op (return audio unchanged).

File: nemo/collections/tts/data/audio_trimming.py [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+
+       Returns integers representing the sample indicies of the start and of speech.
+    """
+    num_frames = is_speech.shape[0]
+
+    # Iterate forwards over the utterance until we find the first frame_threshold consecutive speech frames.
+    start_i = None
+    for i in range(0, num_frames):
+        high_i = min(num_frames, i + frame_threshold)
+        if all(is_speech[i:high_i]):
```
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/data/audio_trimming.py:246]: I think the existing logic works. It can be confusing to follow because for indexing in python the start_i is inclusive and end_i is exclusive.

If i = 1, and frame_threshold = 3, then high_i = 4. Then is_speech[1:4] returns speech frames 1, 2, and 3, which is the correct range.

I removed  the min/max and changed the range instead to avoid checking the boundaries with fewer than "speech_frame_threshold" values.

[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/data/audio_trimming.py:246]: In place of this function there are also utilities in the VAD code (https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/parts/utils/vad_utils.py#L643) which can take a similar input and produce a list with the boundaries for all speech segments. You can then find SoS and EoS by looking at the first and last entries. But I am inclined to try further large refactoring like that to a future PR after we have the general data processing template merged.

File: nemo/collections/tts/data/audio_trimming.py (hunk 4 above)
Comment (rlangman) [reply nemo/collections/tts/data/audio_trimming.py:197]: Added positivity assertions on various input arguments, which also avoids zero-division

File: nemo/collections/tts/data/audio_trimming.py [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+
+        return speech_frames
+
+    def _scale_sample_indices(self, start_sample_i: int, end_sample_i: int, sample_rate: int) -> Tuple[int, int]:
+        sample_rate_ratio = sample_rate / self.vad_sample_rate
+        start_sample_i = sample_rate_ratio * start_sample_i
+        end_sample_i = sample_rate_ratio * end_sample_i
+        return start_sample_i, end_sample_i
+
+    def trim_audio(self, audio: np.array, sample_rate: int, audio_id: str = "") -> Tuple[np.array, int, int]:
```
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/data/audio_trimming.py:202]: Should I add it to each implementation if the docstring is already on the AudioTrimmer(ABC) interface?

File: nemo/collections/tts/data/audio_trimming.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (redoctopus) [reply nemo/collections/tts/data/audio_trimming.py]: New version looks fine to me!

**Review by redoctopus — COMMENTED** [review]
LGTM. @XuesongYang to check over his comments before approval

File: nemo/collections/tts/data/audio_trimming.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (XuesongYang) [reply nemo/collections/tts/data/audio_trimming.py:267]: sounds good!

**Review by XuesongYang — APPROVED** [review]
LGTM. thanks.!
